homeassistant/helpers/script.py

```diff
@@ -444,13 +444,11 @@ async def async_run(self) -> ScriptRunResult | None:
 
         try:
             self._log("Running %s", self._script.running_description)
-            for self._step, self._action in enumerate(self._script.sequence):
-                if self._stop.done():
-                    script_execution_set("cancelled")
-                    break
-                await self._async_step(log_exceptions=False)
-            else:
-                script_execution_set("finished")
+            async with async_interrupt.interrupt(self._stop, ScriptStoppedError, None):
```

Comment:
This will interrupt as soon as the future is set instead of having to wait for it to be checked